### PR TITLE
Support reboot

### LINKED ISSUE #6669: Support reboot
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind enhancement

The current logic for restart is："set termincationGracePeriod and delete the VMI pod to trigger a forced restart"

If we support soft reboot, we can:
1. Support  graceful soft reboot, to allow the VM and filesystem on it to have a chance to shutdown properly.
2. Support reboot button in the Web VNC window.
3. No need to rebuild pod, no pod IP change, no need to reschedule, remount storage, faster restart time.

### PR DESCRIPTION
**Pull request by futuretea**
**What this PR does / why we need it**:
1. expose softreboot(virDomainReboot) to virt-api
```bash
/apis/subresources.kubevirt.io/v1/namespaces/<namespace>/virtualmachineinstances/<name>/softreboot
```
2. add a commad "soft-reboot" to virtctl
```bash
Soft reboot a virtual machine instance

Usage:
  virtctl soft-reboot (VMI) [flags]

Examples:
  # Soft-Reboot a virtualmachineinstance called 'myvmi':
  virtctl soft-reboot myvmi

Flags:
  -h, --help   help for soft-reboot

Use "virtctl options" for a list of global command-line options (applies to all commands).
```
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6669

**Special notes for your reviewer**:

**Release note**:

```release-note
Added 'virtctl soft-reboot' command to reboot the VMI.
```

### CODE REVIEW

File: pkg/virtctl/softreboot/softreboot_test.go
Comment (zcahana) [inline pkg/virtctl/softreboot/softreboot_test.go:9]: ditto about the package alias.

**Review by zcahana — COMMENTED** [review]
Some additional comments.

Also, this new command really needs to be covered by a functional test - especially since it spans many components. For reference and inspiration, you can take a look at [tests/pausing_test.go](https://github.com/kubevirt/kubevirt/blob/main/tests/pausing_test.go)

File: tests/softreboot_test.go
Comment (zcahana) [inline tests/softreboot_test.go:48]: This can be simplified by avoiding the timer+select clause and just using `time.Sleep(30sec)` and then `Except(console.SafeExpectBatch(...))`. Not too much of a big deal though.

**Review by zcahana — APPROVED** [review]
@futuretea I've left one last minor comment, it's not too important and I leave it for your decision whether to change or leave as-is.

Overall - really nice work with this PR, thank you! 

@davidvossel @rmohr @vladikr would one of you care to have a pass and give the final approval?

**Comment by zcahana** [comment]
@futuretea I would also rephrase the release note so that it'd be clearer that a new virtctl command has been added:
```release-note
Added 'virtctl soft-reboot' command to reboot the VM via QEMU guest agent.
```

File: pkg/virt-launcher/virtwrap/manager.go
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/manager.go:1114]: It's unclear to me if we should be talking to the guest agent directly here, or using the virDomainReboot libvirt client command.  It seems like using the virDomainReboot command would be more accurate if multiple methods are possible.

**Review by davidvossel — COMMENTED** [review]
Great feature. Being able to reboot without rescheduling the pod is certainly useful. 

I think there are multiple methods for executing a reboot [1] [2]. Of those methods, the ACPI and guest agent methods seem to be the most relevant ones for us.

I like the idea of using a subresource endpoint like you have done to perform the reboot as an imperative action. Have you thought through how we'd approach supporting multiple reboot methods? Maybe we could pick the best method based on what the VMI has configured?

For example, if the VMI has a guest agent connected (we can see this via the condition on the vmi) then use the guest agent reboot. Otherwise try ACPI if the VMI has ACPI configured. Lastly if no soft reboot method is possible, return an error on the subresource. 

1. https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainRebootFlagValues
2. https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainReboot

File: pkg/virt-launcher/virtwrap/manager.go
Comment (futuretea) [reply pkg/virt-launcher/virtwrap/manager.go]: updated

File: pkg/virt-api/rest/subresource.go
Comment (davidvossel) [inline pkg/virt-api/rest/subresource.go:691]: Added here to the validation, the VMI must either have the agent connected condition or ACPI enabled for soft reboot to be possible.

File: pkg/virt-launcher/virtwrap/manager.go
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/manager.go:1223]: note, it's possible for acpi to be disabled on a VMI.

**Review by davidvossel — COMMENTED** [review]
looks great! i just made one minor comment about needing to validate ACPI is enabled. i think it's technically possible for ACPI to be turned off even though it's on by default

File: pkg/virt-api/rest/subresource.go
Comment (futuretea) [reply pkg/virt-api/rest/subresource.go:696]: updated

File: pkg/virt-launcher/virtwrap/manager.go
Comment (futuretea) [reply pkg/virt-launcher/virtwrap/manager.go]: updated

File: pkg/virt-api/rest/subresource.go
Comment (vasiliy-ul) [inline pkg/virt-api/rest/subresource.go:692]: The `Features` field of the `DomainSpec` struct is optional. Probably it should be null-checked prior to dereferencing it. Same for the `Enabled` of the `FeatureState` (`ACPI` field).

File: pkg/virt-launcher/virtwrap/manager.go
Comment (vasiliy-ul) [inline pkg/virt-launcher/virtwrap/manager.go:1223]: Probably same null-check is needed here: https://github.com/kubevirt/kubevirt/pull/6670/files#r748085148

File: pkg/virt-api/rest/subresource.go
Comment (zcahana) [reply pkg/virt-api/rest/subresource.go:696]: Note that `vmi.Spec.Domain.Features.ACPI.Enabled` is a `*bool` so it would be safer to test it to be non-nil too.

File: pkg/virt-api/rest/subresource.go
Comment (vasiliy-ul) [inline pkg/virt-api/rest/subresource.go:692]: I think the conditions when `vmi.Spec.Domain.Features == nil` and `Spec.Domain.Features.ACPI.Enabled == nil`  should also be checked Maybe smth like below:
```go
if features := vmi.Spec.Domain.Features; features == nil || features.ACPI.Enabled == nil || !(*features.ACPI.Enabled) {
    return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf("VMI neither have the agent connected nor the ACPI feature enabled"))
}
```

File: pkg/virt-launcher/virtwrap/manager.go
Comment (vasiliy-ul) [inline pkg/virt-launcher/virtwrap/manager.go:1223]: Same here
```suggestion
		if features := vmi.Spec.Domain.Features; features == nil || features.ACPI.Enabled == nil || !(*features.ACPI.Enabled) {
```

**Review by vasiliy-ul — COMMENTED** [review]
Still there seems to be a couple of issues with the 'ACPI enabled' conditions. Added comments inline. Hope I havent messed it up :)

File: pkg/virt-api/rest/subresource.go
Comment (zcahana) [reply pkg/virt-api/rest/subresource.go]: @vasiliy-ul re: the above code example, note that ACPI [default to enabled](https://github.com/kubevirt/kubevirt/blob/e622b99f6cfe4cbd659ff408dffb411effa2664c/staging/src/kubevirt.io/client-go/apis/core/v1/schema.go#L1045), so `features == nil || features.ACPI.Enabled == nil` implies ACPI enabled ==> no error.

Comment (vasiliy-ul) [reply pkg/virt-api/rest/subresource.go]: ah... seems that I missed the 'default to enabled' part. so I did mess it up eventually :) sorry, my bad. thanks for noting it.

Comment (futuretea) [reply pkg/virt-api/rest/subresource.go]: updated

**Comment by zcahana** [comment]
/lgtm

Let's wait for a final approval from an approver.
/cc @davidvossel

**Review by davidvossel — COMMENTED** [review]
/approve

thanks!

File: pkg/virt-handler/rest/lifecycle.go
Comment (davidvossel) [inline pkg/virt-handler/rest/lifecycle.go:214]: ah, i missed this, sorry.

If an error occurs, we need to return early. Otherwise we'll send that even that says the reboot was successful.

**Review by davidvossel — COMMENTED** [review]
/approve cancel

File: pkg/virt-handler/rest/lifecycle.go
Comment (futuretea) [reply pkg/virt-handler/rest/lifecycle.go]: updated

**Comment by futuretea** [comment]
@davidvossel updated

**Comment by futuretea** [comment]
@zcahana @davidvossel updated

File: pkg/virt-handler/rest/lifecycle.go
Comment (davidvossel) [inline pkg/virt-handler/rest/lifecycle.go:213]: Everywhere that `WriteError` occurs needs to return early. I think this can technically cause virt-handler to crash otherwise

For example, if socketFile doesn't exist, then NewClient will return nil for the client, and then the client is dereferenced `err = client.SoftRebootVirtualMachine(vmi)`

I know your code is modeled after the pause logic. Whoever wrote that got the same problem wrong. I'll fix that in a separate pr.

Comment (davidvossel) [reply pkg/virt-handler/rest/lifecycle.go:213]: I fix the other areas in  #6841. This new code for reboot needs to follow the same pattern in #6841

**Comment by futuretea** [comment]
@zcahana @davidvossel updated

**Comment by futuretea** [comment]
@zcahana @davidvossel updated

**Comment by zcahana** [comment]
/lgtm

This is definitely good to go in from my perspective. We'll need @davidvossel's final approval though.

**Review by davidvossel — COMMENTED** [review]
/approve

thanks!

**Pull request merged**
